src/server/raft_client.rs

```diff
@@ -11,192 +11,336 @@
+            pool: Builder::new()
+                .name_prefix("raftclient")
+                .pool_size(1)
+                .create(),
+        }
+    }
+
+    fn get_conn(&mut self, region_id: u64, store_id: u64) -> &mut Conn {
+        let index = region_id as usize % self.ctx.cfg.grpc_raft_conn_num;
+        match self.conns.entry((store_id, index)) {
+            Entry::Occupied(e) => e.into_mut(),
+            Entry::Vacant(e) => {
+                let conn = self.ctx.connect(&self.pool, store_id);
+                e.insert(conn)
+            }
         }
     }
 
-    fn get_conn(&mut self, addr: &str, region_id: u64, store_id: u64) -> &mut Conn {
-        let index = region_id as usize % self.cfg.grpc_raft_conn_num;
-        let cfg = &self.cfg;
-        let security_mgr = &self.security_mgr;
-        let env = &self.env;
-        // TODO: avoid to_owned
-        self.conns
-            .entry((addr.to_owned(), index))
-            .or_insert_with(|| Conn::new(Arc::clone(env), addr, cfg, security_mgr, store_id))
+    #[inline]
+    pub fn addr(&mut self, store_id: u64) -> Address {
+        self.ctx.addr(store_id)
     }
 
-    pub fn send(&mut self, store_id: u64, addr: &str, msg: RaftMessage) -> Result<()> {
-        let conn = self.get_conn(addr, msg.region_id, store_id);
-        conn.buffer
-            .as_mut()
-            .unwrap()
-            .push((msg, WriteFlags::default().buffer_hint(true)));
+    // TODO: limit the sending messages.
+    pub fn send(&mut self, store_id: u64, msg: RaftMessage) -> Result<()> {
+        let conn = self.get_conn(msg.region_id, store_id);
+        let msg = (msg, WriteFlags::default().buffer_hint(true));
+        if conn.buffer.len() < MSG_BUFFER_LIMIT {
+            conn.buffer.push_back(msg);
+        } else {
+            // TODO: record metrics.
+            conn.buffer.pop_front();
+            conn.buffer.push_back(msg);
+        }
         Ok(())
     }
 
     pub fn flush(&mut self) {
-        let addrs = &mut self.addrs;
+        let ctx = &mut self.ctx;
+        let pool = &self.pool;
         let mut counter: u64 = 0;
-        self.conns.retain(|&(ref addr, _), conn| {
-            let store_id = conn.store_id;
-            if !conn.alive.load(Ordering::SeqCst) {
-                if let Some(addr_current) = addrs.remove(&store_id) {
-                    if addr_current != *addr {
-                        addrs.insert(store_id, addr_current);
-                    }
-                }
-                return false;
+        for (&(store_id, _), conn) in &mut self.conns {
+            if conn.buffer.is_empty() {
+                continue;
             }
 
-            if conn.buffer.as_ref().unwrap().is_empty() {
-                return true;
-            }
+            ctx.reconnect_if_necessary(pool, conn);
 
             counter += 1;
-            let mut msgs = conn.buffer.take().unwrap();
+            let mut msgs: Vec<_> = conn.buffer.drain(..).collect();
```

Comment:
will `collect` cause a new memory allocation to create the vector?

Reply:
Yes. It's expected.

Reply:
What's the difference? There is still allocation when a message is pushed.